Fix #772: Lowfi Landscape: AppLanguageActivity #772

MohamedMedhat1998 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
-Fixes #772: Lowfi Landscape: AppLanguageActivity   

## Before  
The `AppLanguageActivity` was locked to the portrait mode  
![Screenshot (129)](https://user-images.githubusercontent.com/8533363/77169308-f2070480-6ac1-11ea-8a68-d1fe88c9d321.png)
## After
The `AppLanguageActivity` now supports the landscape mode  
![Screenshot (128)](https://user-images.githubusercontent.com/8533363/77169359-064b0180-6ac2-11ea-81a2-f7cc35678114.png)
## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.
- [x] Implementing a landscape for `AppLanguageActivity`
- [x] Unit tests for `AppLanguageActivity`

MohamedMedhat1998 committed:
-Finished the landscape implementation of "AppLanguageActivity"

MohamedMedhat1998 committed:
-Finished the xml design for the landscape mode for "AppLanguageActivity"

MohamedMedhat1998 committed:
-Finished the test cases for "AppLanguageActivity"

MohamedMedhat1998 commented:
@rt4914 please review this PR

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivity.kt:
```diff
@@ -44,4 +50,9 @@ class AppLanguageActivity : InjectableAppCompatActivity() {
+
+  override fun onSaveInstanceState(outState: Bundle) {
+    super.onSaveInstanceState(outState)
+    outState.putSerializable(KEY_LANGUAGE_SELECTION_ADAPTER ,appLanguageActivityPresenter.getLanguageSelectionAdapter())
```
Spacing is incorrect there should be space after `,` and not before that:
`outState.putSerializable(KEY_LANGUAGE_SELECTION_ADAPTER, appLanguageActivityPresenter.getLanguageSelectionAdapter())`

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivityPresenter.kt:
```diff
@@ -47,4 +47,8 @@ class AppLanguageActivityPresenter @Inject constructor(private val activity: App
+  fun getLanguageSelectionAdapter(): LanguageSelectionAdapter{
+    return languageSelectionAdapter
+  }
+
```
Remove this extra blank line

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivityPresenter.kt:
```diff
@@ -14,10 +14,10 @@ class AppLanguageActivityPresenter @Inject constructor(private val activity: App
   private lateinit var prefSummaryValue: String
 
-  fun handleOnCreate(prefKey: String, prefSummaryValue: String) {
+  fun handleOnCreate(prefKey: String, prefSummaryValue: String,savedAdapter : LanguageSelectionAdapter? = null) {
```
Spacing issue:
`fun handleOnCreate(prefKey: String, prefSummaryValue: String, savedAdapter: LanguageSelectionAdapter? = null) {`

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivityPresenter.kt:
```diff
@@ -47,4 +47,8 @@ class AppLanguageActivityPresenter @Inject constructor(private val activity: App
     languageSelectionAdapter.setDefaultLanguageSelected(prefSummaryValue = prefSummaryValue)
   }
 
+  fun getLanguageSelectionAdapter(): LanguageSelectionAdapter{
```
Spacing issue:
`fun getLanguageSelected(): String {`

rt4914 commented on app/src/sharedTest/java/org/oppia/app/options/OptionsFragmentTest.kt:
```diff
@@ -246,6 +267,30 @@ class OptionsFragmentTest {
+      )
+      onView(withId(R.id.language_recycler_view))
+        .perform(
+          RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(
```
Nit:
`actionOnItemAtPosition<RecyclerView.ViewHolder>(`

rt4914 commented on app/src/main/java/org/oppia/app/options/LanguageSelectionAdapter.kt:
```diff
@@ -7,12 +7,13 @@ import androidx.databinding.library.baseAdapters.BR
 /** Adapter to bind languages to [RecyclerView] inside [AppLanguageActivityPresenter] and [DefaultAudioActivityPresenter]. */
 class LanguageSelectionAdapter(private val prefKey: String) :
-  RecyclerView.Adapter<LanguageSelectionAdapter.LanguageViewHolder>() {
+  RecyclerView.Adapter<LanguageSelectionAdapter.LanguageViewHolder>(),Serializable {
```
We don't want the entire Adapter to be Serializable. Instead check the implementation of `onSavedInstance` in `DefaultAudioActivity`.

rt4914 reviewed (changes requested):
Suggested approach changes.

MohamedMedhat1998 committed:
-Fixed the spacing issues
-Modified the way the "AppLanguageActivity" saves its state on configuration change

rt4914 commented on app/src/main/java/org/oppia/app/options/AppLanguageActivity.kt:
```diff
@@ -19,7 +20,11 @@ class AppLanguageActivity : InjectableAppCompatActivity() {
     activityComponent.inject(this)
     prefKey = intent.getStringExtra(KEY_APP_LANGUAGE_PREFERENCE_TITLE)
-    prefSummaryValue = intent.getStringExtra(KEY_APP_LANGUAGE_PREFERENCE_SUMMARY_VALUE)
+    prefSummaryValue = if(savedInstanceState == null){
```
Spacing issue:
```
    prefSummaryValue = if (savedInstanceState == null) {
      intent.getStringExtra(KEY_APP_LANGUAGE_PREFERENCE_SUMMARY_VALUE)
    } else {
      savedInstanceState.get(KEY_SELECTED_LANGUAGE) as String
    }
    appLanguageActivityPresenter.handleOnCreate(prefKey, prefSummaryValue)
```

rt4914 reviewed (changes requested):
Everything looks great. Just one nit spacing issue change suggested.

MohamedMedhat1998 committed:
-Fixed last spacing issue

rt4914 reviewed (approved):
@MohamedMedhat1998 Everything is perfect in this PR but if you see at the bottom of the PR `non-flaky_tests` are failing and because of we cannot merge this PR. So to fix that we can do following steps:
Merge your branch with latest code from upstream/develop so that your current branch is ahead of `oppia-android/develop` and push that code. Once that is done mostly the CircleCI tests will pass. So that we can merge your PR. Thanks.

MohamedMedhat1998 committed:
Merge branch 'develop' into Lowfi-Landscape-AppLanguageActivity-772

rt4914 commented:
Closing this PR as it has been replicated on #859
